test/Syntax/round_trip_parse_gen.swift

```diff
@@ -532,7 +532,7 @@ struct S : Q, Equatable {
   @_implements(P, x)
   var y: String
   @_implements(P, g())
-  func h() {}
+  func h() { _ = \.self }
```

Comment:
This is a keypath, not a DotSelfExpr. I don't think its correct that DotSelfExpr's base is optional; you should instead parse keypaths differently. CC @jckarter

Reply:
We parse them as DotSelfExpr too for AST, see [here](https://github.com/apple/swift/blob/master/lib/Parse/ParseExpr.cpp#L1165).

Reply:
I see it parses like this:

```
        (keypath_expr type='<null>'
          <<null>>
          (dot_self_expr type='<null>'
            (key_path_dot_expr implicit type='<null>')))))))
```

So the base is a KeyPathDotExpr. Do you not have an equivalent in SwiftSyntax? It would be better if the base was never optional here and with MemberExpr also.

Reply:
Syntax tree doesn't have equivalent implicit nodes in general. All nodes should cover at least one underlying token.

Reply:
The only exception is collection syntax like `CodeBlockItemList`, which can be an empty set of items.

Reply:
`a.self` is creating a `DotSelfExpr` syntax node while `.self` is creating a `ImplicitMemberExpr` one. If we want to be consistent here then shouldn't `\.self` be a `KeyPathExpr` wrapping an `ImplicitMemberExpr` ?

Reply:
Also there is a difference in `.a` creating a `ImplicitMemberExpr` node while `\.a` is a `MemberAccessExpr` wrapped by `KeyPathExpr`.

Reply:
> `a.self` is creating a `DotSelfExpr` syntax node while `.self` is creating a `ImplicitMemberExpr` one. If we want to be consistent here then shouldn't `\.self` be a `KeyPathExpr` wrapping an `ImplicitMemberExpr` ?

After further discussion with Xi, this is explained by the parser treating these differently, `.self` being a `unresolved_member_expr` while `\.self` creating a `dot_self_expr`.

Reply:
yeah, i think this suggestion makes sense. We could reuse existing `MemberAccessExpr` to cover `DotSelfExpr`. We could also use `MemberAccessExpr` to represent `ImplicitMemberExpr` to further reduce the type complexity.